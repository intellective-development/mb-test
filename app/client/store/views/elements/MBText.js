// @flow

import * as React from 'react';
import bindClassNames from '../../../shared/utils/bind_classnames';
import styles from './MBText.scss';

const cn = bindClassNames(styles);

// This module exports a set of wrappers around the basic react text elements
// The idea is to provide a new baseline for us to work off of as we move towards component style sharing.
// Eventually, the goal is to fit these components and props into the styleguide the design team is working towards.

type MBTextProps = { className?: string, reset_spacing?: boolean };

type TextTag = 'p' | 'a' | 'span' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'; // these are the JSX intrinsics (strings). see React.ElementType
export const makeMBText = (TextTagComponent: TextTag, OverrideComponent?: React.ComponentType<*>) => {
  // we use the override component if provided, otherwise we fall back to letting jsx use the built in component that matches the tag string
  const TextComponent = OverrideComponent || TextTagComponent;

  const MBText = ({className, reset_spacing = true, ...rest_props}: MBTextProps) => {
    const classes = cn('elMBText', {elMBText_ResetSpacing: reset_spacing}, className);
    return <TextComponent {...rest_props} className={classes} />;
  };
  MBText.displayName = `MBText(${TextTagComponent})`;

  return MBText;
};

export const Span = makeMBText('span');
export const H1 = makeMBText('h1');
export const H2 = makeMBText('h2');
export const H3 = makeMBText('h3');
export const H4 = makeMBText('h4');
export const H5 = makeMBText('h5');
export const H6 = makeMBText('h6');

// the p component has custom props, used in conjunction with those provided by makeMBText
type PProps = { className: string, body_copy?: boolean };
export const P = makeMBText('p', (({className, body_copy, ...rest_props}: PProps) => {
  const classes = cn('elMBText_P', {elMBText_P__BodyCopy: body_copy}, className);

  return <p {...rest_props} className={classes} />;
}));

// the a component has custom styles, used in conjunction with those provided by makeMBText
type AProps = { className: string, standard?: boolean, body_copy?: boolean };
export const A = makeMBText('a', (({className, standard = true, ...rest_props}: AProps) => {
  const classes = cn('elMBText_A', {elMBText_A__Standard: standard}, className);

  return <a {...rest_props} className={classes} />;
}));
