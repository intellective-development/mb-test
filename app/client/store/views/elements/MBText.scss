@import "../style/variables";

.elMBText {
  font-family: $font-family-avenir;
}

// Foundation adds a bunch of spacing to the bottom of elements
// We allow an override, with a separate class to opt-out
.elMBText_ResetSpacing {
  margin-bottom: initial;
}

.elMBText_P__BodyCopy {
  @include font-size(15);

  color: $brand-medium-grey;
  margin-bottom: 0.75rem;
}

.elMBText_A {
  color: currentColor;

  &:hover,
  &:focus {
    color: currentColor;
  }
}

// we make it the standard red link text optional
.elMBText_A__Standard {
  color: $brand-red;

  &:hover,
  &:focus {
    color: lighten($brand-red, 10%);
  }
}
