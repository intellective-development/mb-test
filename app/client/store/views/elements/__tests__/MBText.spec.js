
import * as React from 'react';

// import TestProvider from 'client/store/__tests__/utils/TestProvider';
import * as MBText from '../MBText';

// tests on the shared HOC behavior

describe('makeMBText', () => {
  const MBTextComponent = MBText.makeMBText('span');
  it('renders', () => {
    expect(MBTextComponent(<MBTextComponent />)).toMatchSnapshot();
  });
});

// tests on the individually exported components

describe('MBText.P', () => {
  it('renders', () => {
    expect(render(<MBText.P />)).toMatchSnapshot();
  });

  it('accepts the body_copy prop', () => {
    expect(render(<MBText.P body_copy />)).toMatchSnapshot();
  });
});

describe('MBText.Span', () => {
  it('renders', () => {
    expect(shallow(<MBText.Span />)).toMatchSnapshot();
  });
});

describe('MBText.H1', () => {
  it('renders', () => {
    expect(shallow(<MBText.H1 />)).toMatchSnapshot();
  });
});

describe('MBText.H2', () => {
  it('renders', () => {
    expect(shallow(<MBText.H2 />)).toMatchSnapshot();
  });
});

describe('MBText.H3', () => {
  it('renders', () => {
    expect(shallow(<MBText.H3 />)).toMatchSnapshot();
  });
});

describe('MBText.H4', () => {
  it('renders', () => {
    expect(shallow(<MBText.H4 />)).toMatchSnapshot();
  });
});

describe('MBText.H5', () => {
  it('renders', () => {
    expect(shallow(<MBText.H5 />)).toMatchSnapshot();
  });
});

describe('MBText.H6', () => {
  it('renders', () => {
    expect(shallow(<MBText.H6 />)).toMatchSnapshot();
  });
});

