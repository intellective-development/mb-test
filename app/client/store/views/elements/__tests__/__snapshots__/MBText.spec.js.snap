// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MBText.H1 renders 1`] = `
<h1
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.H2 renders 1`] = `
<h2
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.H3 renders 1`] = `
<h3
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.H4 renders 1`] = `
<h4
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.H5 renders 1`] = `
<h5
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.H6 renders 1`] = `
<h6
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.P accepts the body_copy prop 1`] = `
<p
  class="elMBText_P elMBText_P__BodyCopy elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.P renders 1`] = `
<p
  class="elMBText_P elMBText elMBText_ResetSpacing"
/>
`;

exports[`MBText.Span renders 1`] = `
<span
  className="elMBText elMBText_ResetSpacing"
/>
`;

exports[`makeMBText renders 1`] = `
<span
  $$typeof={Symbol(react.element)}
  _owner={null}
  _store={Object {}}
  className="elMBText elMBText_ResetSpacing"
  props={Object {}}
  type={[Function]}
/>
`;
